### PR TITLE
Add CLI and batteries to PLDS

### PR DESCRIPTION
**Pull request by philss**
This is the essential to start the project.

Most of the CLI setup was copied from Livebook -
https://github.com/livebook-dev/livebook.

### CODE REVIEW

File: mix.exs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    [
+      {:phoenix, "~> 1.6.0-rc.0", override: true},
+      {:phoenix_live_dashboard, path: "../phoenix_live_dashboard", override: true},
+      {:telemetry_metrics, "~> 0.6"},
+      {:ecto_psql_extras, "~> 0.7"},
+      {:broadway_dashboard, "~> 0.2.1"},
+      {:jason, "~> 1.2"},
+      {:plug_cowboy, "~> 2.5"}
+    ]
+  end
```
Comment (mcrumm) [inline mix.exs:45]: I'm not sure if you're ready to do so yet, or not, but this is what we do to make the LiveView dependency configurable on the dashboard:

```suggestion
      {:phoenix_live_dashboard, phoenix_live_dashboard_opts()},
      {:telemetry_metrics, "~> 0.6"},
      {:ecto_psql_extras, "~> 0.7"},
      {:broadway_dashboard, "~> 0.2.1"},
      {:jason, "~> 1.2"},
      {:plug_cowboy, "~> 2.5"}
    ]
  end

  defp phoenix_live_dashboard_opts do
    if path == System.get_env("LIVE_DASHBOARD_PATH") do
      [path: path, override: true]
    else
      []
    end
  end
```

but it's just a thought! :)

Comment (philss) [reply mix.exs:45]: @mcrumm cool! Thanks! And yeah, I think this is temporary, but can be handy in the future. :smile: 
I will just rebase and add your suggestion after finishing adding some tests.

File: LICENSE.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# MIT License
+
+Copyright (c) 2021 Michael Crumm, Chris McCord, José Valim
```
Comment (josevalim) [inline LICENSE.md:3]:
```suggestion
Copyright (c) 2021 Philip Sampaio
```

:)

File: README.md [hunk 3]
```diff
@@ -1,3 +1,30 @@
-## PLDS
+# PLDS - Phoenix Live Dashboard Standalone
```
Comment (josevalim) [inline README.md:1]:
```suggestion
# PLDS - Phoenix LiveDashboard Standalone
```

File: README.md [hunk 4]
```diff
@@ -1,3 +1,30 @@
...
 
-Work in progress. :)
+PLDS is a bundle of Phoenix Live Dashboard with some tools preinstalled.
+
+It's useful in environments where you can't install Phoenix or the dashboard.
+This can be used to access remote system that are available locally.
```
Comment (josevalim) [inline README.md:6]:
```suggestion
This can be used to access remote system that are reachable locally.
```

File: README.md [hunk 5]
```diff
@@ -1,3 +1,30 @@
...
+Check the options available by typing `plds --help`.
+
+## Preinstalled tools
+
+- Ecto with extras - works for PostgreSQL databases.
+- Broadway Dashboard - to inspect Broadway pipelines.
+
+## License
+
+MIT License. Copyright (c) 2021 Michael Crumm, Chris McCord, José Valim.
```
Comment (josevalim) [inline README.md:30]:
```suggestion
MIT License. Copyright (c) 2021 Philip Sampaio.

File: config/runtime.exs [hunk 6]
```diff
@@ -7,17 +7,9 @@ import Config
 # any compile-time configuration in here, as it won't be applied.
 # The block below contains prod specific runtime configuration.
 if config_env() == :prod do
-  # The secret key base is used to sign/encrypt cookies and other secrets.
-  # A default value is used in config/dev.exs and config/test.exs but you
-  # want to use a different value for prod and you most likely don't want
-  # to check this value into version control, so we use an environment
-  # variable instead.
   secret_key_base =
     System.get_env("SECRET_KEY_BASE") ||
```
Comment (josevalim) [inline config/runtime.exs:11]: Maybe PDLS_SECRET_KEY_BASE? We can mirror the schema livebook uses.

File: lib/plds/application.ex [hunk 7]
```diff
@@ -15,6 +21,55 @@ defmodule PLDS.Application do
...
+
+              elixir --sname test -e "IO.puts node()"
+
+          Then you can try booting PLDS again
+          """)
+      end
+
+      {type, name} = get_node_type_and_name()
+
+      # TODO: Should I start it hidden? If so, how?
```
Comment (josevalim) [inline lib/plds/application.ex:47]: The escript configuration in mix.exs allows you to set emu_args. Pass the -hidden flag there. :)

File: lib/plds_web/endpoint.ex [hunk 8]
```diff
@@ -24,7 +28,7 @@ defmodule PLDSWeb.Endpoint do
     cookie_key: "request_logger"
 
   plug Plug.RequestId
-  plug Plug.Telemetry, event_prefix: [:phoenix, :endpoint]
+  plug Plug.Telemetry, event_prefix: [:plds, :endpoint]
```
Comment (josevalim) [inline lib/plds_web/endpoint.ex:31]: I think we can remove a lot of those plugs. No need for Static, RequestId... maybe not even Plug.Parsers.

File: lib/plds_web/router.ex [hunk 9]
```diff
@@ -4,6 +4,11 @@ defmodule PLDSWeb.Router do
 
   scope "/" do
     pipe_through [:fetch_session, :protect_from_forgery]
-    live_dashboard "/", metrics: PLDSWeb.Telemetry
+
+    get "/health", PLDSWeb.HealthController, :index
+
+    live_dashboard "/",
+      metrics: PLDSWeb.Telemetry,
```
Comment (josevalim) [inline lib/plds_web/router.ex:11]: We can nuke the telemetry stuff I think. Can we set it to false or something to disable the page altogether?

File: mix.exs [hunk 10]
```diff
@@ -20,7 +21,7 @@ defmodule PLDS.MixProject do
   def application do
     [
       mod: {PLDS.Application, []},
-      extra_applications: [:logger, :runtime_tools]
+      extra_applications: [:logger, :runtime_tools, :os_mon, :inets]
```
Comment (josevalim) [inline mix.exs:24]: What do we need inets for? 🤔

Comment (philss) [reply mix.exs:24]: For httpc, to check if plds is already running or the port is in use.

File: lib/plds/application.ex [hunk 11]
```diff
@@ -15,6 +21,24 @@ defmodule PLDS.Application do
     Supervisor.start_link(children, opts)
   end
 
+  # We want to avoid to execute this in test env.
+  if @ensure_distribution do
```
Comment (josevalim) [inline lib/plds/application.ex:25]: What if we keep this as regular application env instead of a compile one?

File: lib/plds/utils.ex [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+    case ip |> String.to_charlist() |> :inet.parse_address() do
+      {:ok, ip} ->
+        ip
+
+      {:error, :einval} ->
+        abort!("expected #{context} to be a valid ipv4 or ipv6 address, got: #{ip}")
+    end
+  end
+
+  @halt_on_abort Application.compile_env!(:plds, :halt_on_abort)
```
Comment (josevalim) [inline lib/plds/utils.ex:24]: Let's make this a regular app env too. :)

File: README.md [hunk 13]
```diff
@@ -1,3 +1,30 @@
...
+
+You can install PLDS using the `mix escript.install hex plds` (or if it's not
+published yet, `mix escript.install github phoenixframework/plds`).
+
+Make sure the path used for installing escripts are available in your
+`PATH` variable.
+
+After that, you can start PLDS with the following command:
+
+    $ plds server --connect node_a,node_b --port 9000 --open
```
Comment (josevalim) [inline README.md:18]: Instead of comma, you can allow the --connect option to be given multiple times by using `:keep` as the type in OptionParser.

**Review by josevalim — APPROVED** [review]
Just three tiny comments and then ship it!!!!

**Pull request merged**